Implement Amplitude-Phase Distance

kiranvad opened the pull request:
Adds amplitude-phase distance following the issue in #43. Computation of similarity metric using the `pipeline.calculate()` is much slower than simply computing the pairwise distance using the apdsit package. Not sure where the bottleneck is.

kiranvad commented:
An example of using the amplitude-phase distance is as follows:
```Python
from AFL.double_agent_tutorial.instruments.tutorial import get_virtual_instrument
from AFL.double_agent.Preprocessor import SAXSLogLogTransform
from AFL.double_agent.Pipeline import Pipeline
from AFL.double_agent.PairMetric import AmplitudePhaseDistance

instrument = get_virtual_instrument()

composition_list = [
    {'a':1/3,'b':1/3,'c':1/3},
    {'a':0.0,'b':0.5,'c':0.5},
    {'a':0.5,'b':0.0,'c':0.5},
    {'a':0.5,'b':0.5,'c':0.0},
]

input_dataset = instrument.measure_multiple(composition_list)

with Pipeline() as test_pipeline:
       SAXSLogLogTransform(
           input_variable='sas',
           output_variable='log_iq',
           dim='q',
        )

result_dataset = test_pipeline.calculate(input_dataset)

with test_pipeline:
    AmplitudePhaseDistance(
        input_variable="log_iq",
        output_variable='similarity',
        sample_dim='log_q',
        )

result_dataset = test_pipeline.calculate(input_dataset)
print(result_dataset) # should have an attribute named similarity of shape (4, 4)
```

martintb commented on AFL/double_agent/Preprocessor.py:
```diff
@@ -52,7 +52,50 @@ def calculate(self, dataset: xr.Dataset) -> Self:
+        input_variable: str = None,
+        output_variable: str = None,
+        dim: str = "q",
+        name: str = "PreprocessorBase",
```
default name should be LogLogTransform

martintb commented on AFL/double_agent/Preprocessor.py:
```diff
@@ -52,7 +52,50 @@ def calculate(self, dataset: xr.Dataset) -> Self:
         """Apply this `PipelineOp` to the supplied `xarray.dataset`"""
         return NotImplementedError(".calculate must be implemented in subclasses")  # type: ignore
 
+class SAXSLogLogTransform(Preprocessor):
```
Maybe rename to LogLogTransform

martintb commented:
Great! A few comments:

1. I think you should revert the domain selection back to the way that I had it. The reason that it worked for me and not for you is that we have opposite understandings on what the `sample` dimension is. You have it set as "log_q" in your example above but it should be whatever your "sample" dimension is called. Here is a typical dataset that we might work with

```python
from AFL.double_agent.data import example_dataset1
example_dataset1() #returns the following
"""
<xarray.Dataset> Size: 85kB
Dimensions:  (sample: 50, q: 100)
Coordinates:
  * q        (q) float64 800B 0.001 0.001072 0.00115 ... 0.8697 0.9326 1.0
Dimensions without coordinates: sample
Data variables:
    I        (sample, q) float64 40kB 1.08e+04 5.558e+03 6.288e+03 ... 1.1 1.002
    dI       (sample, q) float64 40kB 968.7 776.0 629.9 ... 0.09923 0.09923
    model    (sample) <U19 4kB 'broad_peak' 'broad_peak' ... 'mass_fractal'
   """
```
Here you can see that there are `sample` and `q` dimensions associated with the `I` variable. The 'sample' dimension is the one associated with 'samples'. Each scattering pattern is associated with a specific material that is measured so we call that the sample dimension. 

2. Here is an updated test script that changes two things. First, it uses the packaged test data that comes with the repo and, second, it avoids a slight bug in that the test_pipeline was growing with each iteration rather than individually testing the configurations. 

```python

import xarray as xr
xr.set_options(display_expand_data=False)
from AFL.double_agent.Preprocessor import LogLogTransform
from AFL.double_agent.Pipeline import Pipeline
from AFL.double_agent import *
from AFL.double_agent.AmplitudePhaseDistance import AmplitudePhaseDistance
import numpy as np 
import time 

input_dataset = example_dataset2()
print(input_dataset)

# test log-log preprocessor
with Pipeline() as test_pipeline:
       LogLogTransform(
           input_variable='I',
           output_variable='log_iq',
           dim='q',
        )

# test apdist pairmetric
for method in ["discrete", "continuous"]:
    print("testing for %s..."%method)
    start = time.time()
    with test_pipeline.copy() as tp:
        AmplitudePhaseDistance(
            input_variable="log_iq",
            output_variable='similarity',
            method=method,
            sample_dim='sample',
            )

    result_dataset = tp.calculate(input_dataset)
    D = result_dataset.similarity.values
    print(D)
    assert np.isfinite(D).all(), "Distances are not finite"
    assert (D>=0).all(), "Distances are not positive"
    end = time.time()
    print("total time elapsed : ", end-start)
```
    
3. I'm getting a bug in apdist during the "discrete" test:
 
```
    File "/home/codespace/.python/current/lib/python3.12/site-packages/apdist/distances.py", line 55, in _phase_distance
    theta = np.trapezoid(np.sqrt(gam_dev), x=time)
            ^^^^^^^^^^^^
  File "/home/codespace/.local/lib/python3.12/site-packages/numpy/__init__.py", line 333, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'trapezoid'
```
did you want `np.trapz` instead? 

4.  I'm also getting this error for the continuous version

```
Traceback (most recent call last):
  File "/workspaces/AFL-agent/test.py", line 33, in <module>
    result_dataset = tp.calculate(input_dataset)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/AFL-agent/AFL/double_agent/Pipeline.py", line 392, in calculate
    op.calculate(dataset1)
  File "/workspaces/AFL-agent/AFL/double_agent/AmplitudePhaseDistance.py", line 120, in calculate
    pair_dists = scipy.spatial.distance.pdist(codomain, metric=metric)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/codespace/.local/lib/python3.12/site-packages/scipy/spatial/distance.py", line 2315, in pdist
    return _pdist_callable(X, metric=metric, out=out, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/codespace/.local/lib/python3.12/site-packages/scipy/spatial/distance.py", line 2771, in _pdist_callable
    dm[k] = metric(X[i], X[j], **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/AFL-agent/AFL/double_agent/AmplitudePhaseDistance.py", line 112, in <lambda>
    metric = lambda y1, y2 : self._get_pairiwise_ap(self.method, 
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/AFL-agent/AFL/double_agent/AmplitudePhaseDistance.py", line 133, in _get_pairiwise_ap
    amplitude, phase, _ = torch_apdist(torch.from_numpy(t),
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/codespace/.python/current/lib/python3.12/site-packages/apdist/torch.py", line 92, in TorchAmplitudePhaseDistance
    f1 = Function(t, f1.reshape(-1,1))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/codespace/.python/current/lib/python3.12/site-packages/funcshape/functions.py", line 21, in __init__
    coeffs = natural_cubic_spline_coeffs(self.x, self.fx)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/codespace/.python/current/lib/python3.12/site-packages/torchcubicspline/interpolate.py", line 217, in natural_cubic_spline_coeffs
    _validate_input(t, x)
  File "/home/codespace/.python/current/lib/python3.12/site-packages/torchcubicspline/interpolate.py", line 9, in _validate_input
    raise ValueError("t must both be floating point.")
ValueError: t must both be floating point.
```

kiranvad commented:
Thanks! I now understand how you're obtaining the domain values for the `I(q)` variables, so I’ve adjusted that accordingly in the PR.

The NumPy error was due to the version being older than 2.0. I've updated the `apdist` repo to reflect the required changes—reinstalling the package from the GitHub repo should resolve it.

Regarding the error from the continuous version, I wasn’t able to reproduce it locally. I believe it may have occurred because the domain values passed in were integer sample IDs instead of floats. The latest commit in this PR should address that issue.

Let me know if you still encounter it!

martintb commented:
This all looks great. Ideally I'd like a couple of tests but I think we can add them later if we want to now that the test suite is merged into main. The only last change I request is for you to make two small changes to pyproject.toml. I'll merge after that. 

Please add the following above the [project] section:
```toml
[tool.hatch.metadata]
allow-direct-references=true
```

And the following somewhere under the [project.optional-depedencies] section:
```toml
apdist = [
        "apdist @ git+https://github.com/kiranvad/Amplitude-Phase-Distance"
]
```
